README.md
# h2buster (v0.4d) #
A fast, threaded, recursive, web directory brute-force scanner over HTTP/2 using [hyper](https://github.com/Lukasa/hyper), inspired by [Gobuster](https://github.com/OJ/gobuster).

## Features ##
* Fast and portable - install [hyper](https://github.com/Lukasa/hyper) and run.
* Multiconnection scanning.
* Multithreaded connections.
* Scalable: scans can be as docile or aggressive as you configure them to be.
* h2 and h2c support.
* Configurable directory recursion depth.
* robots.txt scanning.
* Multiplatform: works on both \*nix and Windows.

## Install ##
Requires Python 3.6.
You only need to install one dependency. If you don't have [hyper](https://github.com/Lukasa/hyper), run:\
`pip3 install -r requirements.txt`

## Usage ##
```
usage: h2buster.py [-h] -w wordlist -u target [-c connections=8]
                   [-t threads=20] [-m http_method=HEAD]
                   [-r directory_depth=2] [-hd header_list]
                   [-x extension_list] [-b http_code_list] [-l] [-vr] [-wc]
                   [-rb] [-nc]

h2buster: an HTTP/2 web directory brute-force scanner.

arguments:
  -h, --help            show this help message and exit
  -w wordlist           Directory wordlist
  -u target             Target URL/IP address ([scheme://]host[:port]).
                        Default port is 443 and HTTPS enabled. To specify
                        otherwise, use ':port' and/or 'http://' (port will
                        default to 80 then).
  -c connections=8      Number of HTTP/2 connections.
  -t threads=20         Number of threads per connection.
  -m http_method=HEAD   HTTP request method. Allowed values are GET, HEAD.
  -r directory_depth=2  Maximum recursive directory depth. Minimum is 1,
                        unlimited is 0.
  -hd header_list       List of headers in the format
                        'header->value|header->value...'. For example: -hd
                        'user-agent->Mozilla/5.0|accept-encoding->gzip,
                        deflate, br'.
  -x extension_list     List of file extensions to check separated by a
                        vertical bar (|). For example, -x '.php|.js|blank|/'.
                        The 'blank' keyword signifies no file extension.
                        Default extensions are '/', 'blank', '.html', '.php'
  -b http_code_list     List of blacklisted response codes separated by a
                        vertical bar (|). Directories with these response
                        codes will not be shown in the output. Default is 404.
  -l                    Flag: show response length in output. This overrides
                        the request method to GET.
  -vr                   Flag: force TLS certificate verification.
  -wc                   Flag: request a random path and analyze response to
                        detect false positives (wildcard processing).
  -rb                   Flag: scan for a robots.txt file. If found, a prompt
                        will be displayed asking whether to use the results.
  -nc                   Flag: disable colored output text.
```

### Example ###
```
$ python3 h2buster.py -w test/small.txt -u www.google.com -c4 -t15 -m GET -x 'blank|/' -r1 -hd 'user-agent->h2buster' -b '404|301'
```
This will scan `www.google.com`, over 4 connections with 15 threads each, performing `GET` requests over the input wordlist, `small.txt`. Each entry in the wordlist will be sent as-is, as well as with a slash (`/`) appended to it. Found directories will not be scanned recursively (depth is 1). The string `h2buster` will be sent as an user agent. The response codes 404 and 301 will not be shown in the output.

## Contributing ##
Check the [TODO](TODO.md) file for a list of features that need work.


CHANGELOG.md
# Changelog #

## 0.4d ##
* Added an option to perform wildcard response detection (`-wc`). This way, if a server replies with codes different than 404 for resources that are not found (for example with a redirection), those can be detected as well.
* The input target is first checked for redirections. If a redirect is detected, the user is asked whether to update the target to the redirection location or not.
* Changed the default number of connections to the number of cores in the target system multiplied by 2.
* Improved cross-compatibility for Unix-like systems when spawning processes.
* Entries found that are identified as directories are now indicated as such in the output.
* Improved error and interrupt handling.
* Minor performance enhancements.

## 0.4c ##
* Added an option to change the HTTP request method (`-m`). Valid values are HEAD (default) and GET.
* Added an option to display the response size (`-l`). This overrides the request method to GET.
* robots.txt scan output now more verbose.
* Fixed the case where visiting /robots.txt gives a redirection (when using `-rb`).
* Minor performance improvements.

## 0.4b ##
* Several performance improvements.
* Bug fixes in the robots.txt parser.
* Sitemaps in robots.txt are now reported for manual inspection if found (they are not parsed).
* Added an option (`-vr`) to verify TLS certificates (otherwise they are not checked).
* Duplicated entries are properly filtered, now taking into account file extensions.
* Status code display made cleaner.

## 0.4a ##
* Updated most strings to f-strings. **This makes Python 3.6 a requirement**.
* Added the option to scan for the robots.txt file (`-rb`):
	- If found, the user is prompted about whether to use its information or not.
	- The user can either retrieve all entries in the file, just the allowed ones based on our own User Agent, or completely ignore the file.
	- HOWEVER, dictionary entries **ARE NOT CHECKED** against the robots.txt rules. Use your wordlist at your own risk. I might add the option to only use dictionary entries if they are allowed as a command line option in the future.
	- The information obtained from this file can be used in a smarter way. For now all the directories found are checked in their respective recursive depth. The entry `/a/b/c` will result in checking `/a` in the first iteration. If `/a` is found, `/a/b` will be searched in the next recursive iteration. Repeat this process for `/a/b/c`.
* Reset HTTP/2 streams:
	- More information is given about the process which handles that stream.
	- Increased sleep time for the thread handling that reset.
* Removed an unnecessary include. The rest of the includes are now tidier.
* Duplicated entries in the input wordlist are just requested once.
* Increased modularity by moving parser algorithms to external classes.
* Removed the `enable_push` parameter for a call to the underlying hyper library - some versions don't seem to accept it.
* Changed the way time is benchmarked. Now it represents how much seconds the actual scan took (as opposed to the time of option parsing & checking + the scan).

## 0.3f ##
* Added an option to ignore specific response codes (`-b`) by providing a list of codes separated by a vertical bar (`|`). Default is 404.

## 0.3e-2 ##
* Improved error handling:
	- Now processes exit gracefully when things go wrong in the middle of a scan instead of hanging.
	- Keyboard interrupt is now less ugly.
* Changed default connections (`-c`) to 4. This seems to yield a performance improvement in most cases.
* Changed `--help` text to be tidier.
* Changed line endings to UNIX-style (in case you were trying to run as `./h2buster.py`).

## 0.3e-1 ##
* Improved error handling for non-RFC-compliant HTTP/2 servers.

## 0.3e ##
* A list of headers can be given to be sent for each request with `-hd` with the format `-hd 'header->value[|header->value|header->value...]'`. For example: `-hd 'user-agent->Mozilla/5.0|accept-encoding->gzip, deflate, br'`.
* Extensions are now separated by a vertical bar too (`|`) for consistency (e.g. `-x '.php|.js|blank|/'`).
* The `server` header of the first response is now displayed at the beginning of the scan (if there is one).

## 0.3d-1 ##
* Improved error handling for reset connections, HTTP/1-only targets, targets that do not exist and TLS errors.

## 0.3d ##
* A list of extensions can be given to be scanned, separated by a semicolon, with `-x`. For example, `-x '.php;.js;blank;/'` will check for .php, .js, blank and / file endings. Note that the `blank` keyword is used to signify no file ending.
* Improved target parsing (`-u`).
* Added feedback on stdout to see current entry being scanned (only on Linux and OS X).
* Changed default threads (`-t`) from 15 to 20.
* Improved color printing performance. The program should run smoother on both UNIX-based and Windows.
* Other very slight performance improvements.

## 0.3c-1 ##
* Fixed a bug where not using `-nc` on Windows would crash the program.
* Fixed a bug where found directories with a space on them would not be properly URL encoded when scanning them.

## 0.3c ##
* HTTP/2 support on the target is checked before starting scan. If the target is not compatible or resets connection, the program exits more gracefully.
* Added colors in output (only for Linux and OS X). To disable them, use `-nc`.

## 0.3b ##
* Fixed a bug where for some entries in the wordlist would crash the script.
* Improved directory detection for found entries:
	- Some found URLs ending in `/` are no longer classified as directories (such as `/index.php/`).
	- URLs ending in `//` are no longer considered directories.

## 0.3a ##
* Added multiprocessing. Each process handles its own HTTP/2 connection. Number of connections can be specified with `-c`. This should greatly improve speed.

lib2hbuster.py
import fnmatch

class RobotParser:

	def __init__(self, user_agent=""):
		self.user_agent = user_agent
		self.entries = set()

	def parse(self, content, policy="allow"):
		policy = policy.lower()
		if policy not in ["all",  "allow"]:
			raise ValueError("Invalid robot policy.")

		# Clean previous entries
		self.entries = set()

		content = content.split("\n")
		for line in content:
			if len(line) == 0: continue
			if line[0] == "#": continue
			if line.count(":") == 0: continue

			line = [x.lstrip().split("$")[0] for x in line.rstrip().lower().split(":", 1)]

			if line[0] == "user-agent":
				if policy=="allow" and fnmatch.filter(self.user_agent, line[1]):
					applies = True
				else:
					applies = False

			elif line[0] in ["allow", "disallow"]:
				if applies == False or (applies == True and line[0] == "allow"):
					if len(line[1])>0:	self.entries.add(line[1])

	def get_entries(self):
		return self.entries

class UrlParser:

	def __init__(self, default_secure=True, default_port=443):
		self.default_secure = default_secure
		self.default_port = default_port

	def parse(self, url):
		# Remove GET parameters
		if url.count("?") > 0:
			target, params = tuple(url.split("?", 1))
		else:
			target = url
			params = ""

		# Protocol
		target = target.split("://", 1)
		if len(target) == 1:
			url = target[0]
			s = None
		else:
			target[0] = target[0].lower()
			url = target[1]
			if target[0] == "http": s = False
			elif target[0] == "https": s = True
			else: raise ValueError("Invalid URL: protocol scheme not recognized.")
			
		# Directory 
		url = url.split("/", 1)
		tup = url[0]
		if len(url) == 1 or len(url[1])==0:
			directory = "/"
		else:
			directory = "/" + url[1]
			if directory[-1]!="/": directory = directory + "/"

		# IP / port
		if tup.count(":") == 0:
			ip = tup
			if s == True: port = 443
			elif s == False: port = 80
			else:
				s = self.default_secure
				port = self.default_port
		elif tup.count(":") == 1:
			try:
				ip, port = tup.split(":", 1)
				port = int(port)
			except ValueError: raise ValueError("Invalid URL: Non-numeric port.")

			if port == 80 and target[0]!="https": s = False
			else: s = True
			
		else:
			raise ValueError("Invalid URL: parsing error.")

		# Save values to object properties
		self.secure = s
		self.host = ip
		self.port = port
		self.path = directory
		self.params = params

	def __repr__(self):
		x = f"TLS: {self.secure}"
		x = f"{x}\nHost: {self.host}"
		x = f"{x}\nPort: {self.port}"
		x = f"{x}\nPath: {self.path}"
		x = f"{x}\nParams: {self.params}"
		return x


libh2buster.py
import fnmatch
import random, string

class NotfoundWildcard:

	__slots__ = (
		"request_string",
		"status",
		"redirect",
		"real_redirect",
		"content",
		"real_content",
		"response_length"
	)

	def __init__(self, connection, method, redirect_codes):
		self.request_string = "/" + "".join((random.choice(string.ascii_lowercase) for _ in range(20)))
		sid = connection.request(method, self.request_string)
		resp = connection.get_response(sid)

		if resp.status in redirect_codes:
			self.redirect = resp.headers.get(b'location')[0].decode('utf-8')
			self.real_redirect = self.redirect.replace(self.request_string, "")
		else:
			self.redirect = False
			self.real_redirect = False

		self.content = resp.read().decode('utf-8')
		self.real_content = self.content.replace(self.request_string, "")
		self.response_length = len(self.content)
		self.status = resp.status

	def match(self, request_string, content, status, redirect):
		if status == 404:
			return False

		if self.redirect and redirect:
			return (self.real_redirect == redirect.replace(request_string, ""))

		elif self.response_length > 0 and len(content) > 0:
			return (self.real_content == content.replace(request_string, ""))

		else:
			return False

class RobotParser:

	def __init__(self, user_agent=""):
		self.user_agent = user_agent
		self.entries = set()
		self.sitemaps = set()

	def parse(self, content, policy="allow"):
		policy = policy.lower()
		if policy not in ("all",  "allow"):
			raise ValueError("Invalid robot policy.")

		# Initialize variables for this parsing call
		self.entries = set()
		self.sitemaps = set()
		disallowed = set()
		applies = False
		content = content.split("\n")

		# Loop over robots.txt content
		for line in content:
			if len(line) == 0 or line[0] == "#" or line.count(":") == 0: continue
			line = tuple([x.lstrip().split("$")[0] for x in line.rstrip().lower().split(":", 1)])

			# Rules for a specific useragent
			if line[0] == "user-agent":
				if policy=="allow" and fnmatch.filter(self.user_agent, line[1]):
					applies = True
				else:
					applies = False

			# Entry for a specific URI
			elif line[0] in ("allow", "disallow"):
				if (not applies and line[1] not in disallowed) or (applies and line[0] == "allow"):
					if len(line[1])>0:
						self.entries.add(line[1])

				elif applies and line[0] == "disallow":
					disallowed.add(line[1])
					self.entries.discard(line[1])

			# URL for a sitemap
			elif line[0] == "sitemap":
				self.sitemaps.add(line[1])

	def get_entries(self):
		return self.entries

	def get_sitemaps(self):
		return self.sitemaps


class UrlParser:

	def __init__(self, default_secure=True, default_port=443):
		self.default_secure = default_secure
		self.default_port = default_port

	def parse(self, url):

		# Remove GET parameters
		if url.count("?") > 0:
			target, params = url.split("?", 1)
		else:
			target = url
			params = ""

		# Protocol
		target = target.split("://", 1)
		if len(target) == 1:
			url = target[0]
			s = None
		else:
			target[0] = target[0].lower()
			url = target[1]
			if target[0] == "http": s = False
			elif target[0] == "https": s = True
			else: raise ValueError("Invalid URL: protocol scheme not recognized.")
			
		# Directory 
		url = url.split("/", 1)
		tup = url[0]
		if len(url) == 1 or len(url[1])==0:
			directory = "/"
		else:
			directory = "/" + url[1]
			if directory[-1]!="/": directory = directory + "/"	# force URIs to end in '/'

		# IP / port
		if tup.count(":") == 0:
			ip = tup
			if s == True: port = 443
			elif s == False: port = 80
			else:
				s = self.default_secure
				port = self.default_port
		elif tup.count(":") == 1:
			try:
				ip, port = tup.split(":", 1)
				port = int(port)
			except ValueError: raise ValueError("Invalid URL: Non-numeric port.")

			if port == 80 and target[0]!="https": s = False
			else: s = True
			
		else:
			raise ValueError("Invalid URL: parsing error.")

		# Save values to object properties
		self.secure = s
		self.host = ip
		self.port = port
		self.path = directory
		self.params = params

h2ws.py
#coding=utf-8

__author__ = "0xc"
__version__ = "0.1b"

try:
	import hyper
	import ssl, sys, time

	# Parse target and check if its HTTPS
	def check_tls(ip):

		ip = ip.split("://")

		if len(ip) == 1:
			s = 0
			ip = ip[0]
		else:
			if ip[0] == "http": s = 0
			elif ip[0] == "https": s = 1
			else: sys.exit("[-] Target not understood")
			ip = ip[1]

		return s, ip

	# Connect to target and return connection object
	def h2_connect(ip, s):

		# Get port
		if len(ip.split(":")) == 1:
			if s==1: port=443
			elif s==0: port=80
		else: port = int(ip.split(":")[-1])

		# Perform connection
		if s == 1:
			ctx = ssl.SSLContext()
			ctx.set_alpn_protocols(['h2'])
			ctx.verify_mode = ssl.CERT_NONE
			conn = hyper.HTTP20Connection(ip, port=port, ssl_context=ctx, enable_push=False)
		elif s == 0:
			conn = hyper.HTTP20Connection(ip, port=port, enable_push=False)

		# Test connectivity before starting the scan
		conn.ping("00000000")

		return conn

	# Read responses, print results and return found directories
	def dump_scan(requests):

		# List of directories that didn't return 404
		o = list()

		for key, element in requests.items():
			resp = conn.get_response(element)
			status = resp.status
			#resp.read(decode_content=True) # This seems to improve performance
			resp.close()

			# Print meaningful results
			if status != 404:
				if key[-1] == "/":
					o.append(key)
					ftype = " <DIR>"
				else: ftype = ""
				print("[+] " + key + ": " + str(status) + ftype)

				# Print redirection location?
				#if status == 301: print(resp.headers)
		return o

	# Scan directory over connection "conn" with dictionary "file". Admits recursion.
	def recursive_dirscan(conn, directory, file, ext):

		global total_length

		i = 0
		requests = dict()
		found = list()

		l = ["|", "/", "-", "\\"]
		t = 0

		# Main loop
		with open(file, "r") as f:
			for entry in f:
				entry = entry.rstrip()

				if entry == "/" or entry=="": continue

				for ex in ext:

					# Prevent flood
					time.sleep(0.1)

					# Rotating bar
					if int(time.time())%5==0:
						print(l[t], end="\r")
						t = (t+1)%4

					# This controls how often we read responses and update results
					# If this value is too high we're opening too many streams without reading responses
					# If this value is too low we're reading too often (not using stream multiplexing effectively due to single thread)
					# This value (20) is eyeballed, it should be tested more thoroughly
					if i>20:
						found += dump_scan(requests)
						requests.clear()
						i = 0

					# Don't need to scan entry.php/ so we skip it
					if len(entry.split(".")[-1]) in ext and ex=="/": continue

					# Perform request and store stream ID
					sid = conn.request("HEAD", directory + entry + ex)
					requests[directory + entry + ex] = sid
					i += 1

			found += dump_scan(requests)

			# Recursively scan found directories
			for fd in found:
				print(" \n[*] Scanning " + fd)
				recursive_dirscan(conn, fd, file, ext)

	if __name__ == "__main__":

		# For benchmarking purposes
		t0 = time.time()

		# Some basic input checking
		if len(sys.argv)<3:
			print("Usage: " + sys.argv[0] + " <dictionary> <target>")
			sys.exit()
		file = sys.argv[1]
		ip = sys.argv[2]

		# For every entry in the dictionary, every extension will be checked
		ext = ["/", ".php", ".html", ".htm", ".asp", ".js", ".css"]

		# Check HTTP/HTTPS
		s, ip = check_tls(ip)

		# Open connection
		conn = h2_connect(ip, s)
		print("[*] Starting scan on " + ip)

		# Main function
		recursive_dirscan(conn, "/", file, ext)

		print("\n[*] Program ran in " + str(round(time.time()-t0, 3)) + " seconds")

except ImportError:
	print("[-] hyper missing: pip3 install hyper")

except OSError as ose:
	if ose.errno == 113: print("[-] No route to host")
	else: print(ose)

except FileNotFoundError:
	print("[-] File not found")

except AssertionError:
	print("[-] H2 not supported for that target")

except KeyboardInterrupt:
	print("\r[-] Scan aborted")